code/game/objects/items/stacks/stack.dm

```diff
@@ -133,145 +133,154 @@
 
 /obj/item/stack/proc/get_amount()
 	if(is_cyborg)
-		. = round(source.energy / cost)
+		. = round(source?.energy / cost)
 	else
 		. = (amount)
 
+/**
+  * Builds all recipes in a given recipe list and returns an association list containing them
+  *
+  * Arguments:
+  * * recipe_to_iterate - The list of recipes we are using to build recipes
+  */
+/obj/item/stack/proc/recursively_build_recipes(list/recipe_to_iterate)
+	var/list/L = list()
+	for(var/recipe in recipe_to_iterate)
+		if(istype(recipe, /datum/stack_recipe_list))
+			var/datum/stack_recipe_list/R = recipe
+			L["[R.title]"] = recursively_build_recipes(R.recipes)
+		if(istype(recipe, /datum/stack_recipe))
+			var/datum/stack_recipe/R = recipe
+			L["[R.title]"] = build_recipe(R)
+	return L
+
+
+/**
+  * Returns a list of properties of a given recipe
+  *
+  * Arguments:
+  * * R - The stack recipe we are using to get a list of properties
+  */
+/obj/item/stack/proc/build_recipe(datum/stack_recipe/R)
+	return list(
+		"res_amount" = R.res_amount,
+		"max_res_amount" = R.max_res_amount,
+		"req_amount" = R.req_amount,
+		"ref" = "\ref[R]",
+	)
+
+/**
+  * Checks if the recipe is valid to be used
+  *
+  * Arguments:
+  * * R - The stack recipe we are checking if it is valid
+  * * recipe_list - The list of recipes we are using to check the given recipe
+  */
+/obj/item/stack/proc/is_valid_recipe(datum/stack_recipe/R, list/recipe_list)
+	for(var/S in recipe_list)
+		if(S == R)
+			return TRUE
+		if(istype(S, /datum/stack_recipe_list))
+			var/datum/stack_recipe_list/L = S
+			if(is_valid_recipe(R, L.recipes))
+				return TRUE
+	return FALSE
+
 /obj/item/stack/attack_self(mob/user)
 	interact(user)
 
-/obj/item/stack/interact(mob/user, sublist)
-	ui_interact(user, sublist)
+/obj/item/stack/interact(mob/user)
+	ui_interact(user)
+
+/obj/item/stack/ui_state(mob/user)
+	return GLOB.hands_state
+
+/obj/item/stack/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Stack", name)
+		ui.open()
+
+/obj/item/stack/ui_data(mob/user)
+	var/list/data = list()
+	data["amount"] = get_amount()
+	return data
+
+/obj/item/stack/ui_static_data(mob/user)
+	var/list/data = list()
+	data["recipes"] = recursively_build_recipes(recipes)
```

Comment:
You don't really need to rebuild this on every open for every panel, store the result and do it only once,

Reply:
If I understood that correctly - I could make a list variable on the stack item, and fill that with recipes by using recursively_build_recipes proc in Initialize(), and then simply use that list in here?

Reply:
Then it would be recalculated on every stack splitting, which is also not great.

Reply:
Yes, but it should be a static/global since it's same for every stack of given type.

Reply:
Having a little problem with this. I have tried both style's and mine solution regarding this, but outcome was the same - static list was shared between ALL subtypes and was not contained just to its own type, so glass stack had a metal recipes and so forth.

Reply:
Alright then, keep the current setup.